### ISSUE TITLE
Revisit EdgeResponseHandler

### ISSUE THREAD

**Issue by emdobrin** [adobe/aepsdk-edge-ios#90]
As we start having more events which return event handles back to the client, we should evaluate if the EdgeResponseHandler is fulfilling the needs. Discuss about onError, onComplete callback.
Evaluate if a regular completion handler can be used instead of a protocol.

**Comment by shalehaha** [comment]
Prefer using a closure instead of a protocol. 
Definitely need  onError, onComplete callback, or we can combine all the things into one closure, something like:
```
(data, error, isLast ) -> Void
```

Another thing I want to discuss, with today's API,`EdgeResponseHandler` is designed to be called multiple times. It is not easy for a client (customer or other extension built on top of edge) to handle the responses this way, it is possible for edge to wait for all the responses and then return them as an array to the caller?

**Comment by shalehaha** [comment]
```
public protocol EdgeResponseHandler {
    func onPartialResponse(data: [String: Any])
    func onComplete()
    func onErrors(errors:[EdgeEventError])
}
```
@emdobrin according to our discussion before, is this the correct way how we want to change the `EdgeResponseHandler`?

Do we need to differentiate between errors vs warnings?

**Comment by emdobrin** [comment]
@shalehaha here are the two options I think that cover all the use-cases you added in here:

1. Incremental updates, get advantage of the streaming option and get updated right away as we continue to process the data: 
```
public protocol EdgeResponseHandler {
    func onResponseUpdate(eventHandle: EdgeEventHandle) // may be called multiple times
    func onErrorUpdate(error: EdgeEventError) // may be called multiple times
    func onComplete() // called when the processing is over, onResponseUpdate and onErrorUpdate will not be called anymore after this
}
static func sendEvent(experienceEvent: ExperienceEvent, responseHandler: EdgeResponseHandler? = nil) // unchanged

Usage:
Edge.sendEvent(experienceEvent: experienceEvent, responseHandler: DemoResponseHandler())
```
2. Have a completion closure which returns all the handles / errors for the corresponding event as lists once the Edge extension finished processing the server response.
```
static func sendEvent(experienceEvent: ExperienceEvent, completion: (([EdgeEventHandle], [EdgeEventError]) -> Void)? = nil) {
    }

Usage:
Edge.sendEvent(experienceEvent: experienceEvent) { (handles: [EdgeEventHandle], errors: [EdgeEventError]) in
     //code
}
```
The second option requires some additional changes on the way we handle the responses today as we need to queue the handles/errors and slightly increases the memory on our side (we need to evaluate if this would be a real concern).

Both of these options assume that we make the `EdgeEventHandle` and `EdgeEventError` public which we should discuss. The alternative is that we pass the handles and errors as dictionaries [String: Any], which might be a bit more inconvenient to use. 

If you need the information about the error type (error/warning) we can add a field in EdgeEventError, something like `errorLevel: error/warning`.
Please note that EdgeEventError will change with https://github.com/adobe/aepsdk-edge-ios/issues/91

Is the event uuid still required for your use-case?

**Comment by shalehaha** [comment]
Should Edge extension provide both?
I'm thinking option #1 works well naturally with the way how Edge returns responses, and in the future customer/extension can build some advanced features around it. 
But #2 is easier to use, especially in the scenario where the caller doesn't care much about the speed and knows there will only be one response block for it. And we can easily build #2 on top of the API provided by #1.

> If you need the information about the error type (error/warning) we can add a field in EdgeEventError, something like errorLevel: error/warning.
My question is more about, for a normal scenario, does the caller need to handle the warnings differently than errors? From what I have seen, seems like they both mean there is something wrong on the backend. So wondering why Edge network treats them differently.

> Is the event uuid still required for your use-case?
Yes, we will need the event uuid.

**Comment by emdobrin** [comment]
@shalehaha regarding 
> does the caller need to handle the warnings differently than errors? 

As a caller I think you will handle them similarly, I don't know if this extra information helps in any way, you can handle them based on the error code if needed. We currently use the error type as log level.

**Comment by emdobrin** [comment]
I would prefer keeping only one option, whichever is more ergonomic for the developer, but maybe we can keep both for beta and get to play with them. Option 1 is also available today by registering event listeners, you can listen to event handles, or errors or both. 
For making the EdgeEventError and EdgeEventHandle public, these classes are mapped on the server response format which may change in the future (e.g. new version of the endpoint) and may cause API deprecation. As an extension developer it is easier to identify these changes when using a class, but having it as dictionary allows us to have all params as optionals. @shalehaha any preferences?

**Comment by shalehaha** [comment]
> Option 1 is also available today by registering event listeners, you can listen to event handles, or errors or both.

As we discussed last time, this is not true today. 1) Edge extension doesn't return event id, and without it, we couldn't listen for event handles or errors. 2) Core today doesn't have an API to allow you to listen for a random event with a certain response id.

Also, I prefer option 2, it is just simple to use.

For EdgeEventError and EdgeEventHandle, I think they can be classes as long as we defined the fields in a general way. That should tolerate minor changes on the server sides. But if there is any significant change in the server response, then we have to update all the extensions either way.
